File: apps/api/app/schemas/equine.py
from datetime import date, datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, Field

from app.common.enums import (
    EquineExperienceFit,
    EquineLocationStatus,
    EquineOperationalStatus,
    EquineSex,
    EquineSpecies,
)
from app.schemas.common import AuditMetadataSchema


class EquineCreateSchema(BaseModel):
    # Identificación base / CV
    name: str = Field(min_length=1, max_length=120)
    inventory_number: int | None = Field(default=None, ge=1)

    # Equivalente a CLASE: MUL. / ASN. / CAB.
    species: EquineSpecies = EquineSpecies.MULE

    # Equivalente a UBICACIÓN: LA JUANA / OTROS
    location_status: EquineLocationStatus = EquineLocationStatus.LA_JUANA
    location_notes: str | None = None

    # Datos zootécnicos / ficha
    breed: str | None = None                 # RAZA real: Criolla, Lusitana, Criolla x Lusitana
    sex: EquineSex = EquineSex.UNKNOWN
    coat_color: str | None = None            # COLOR
    gait: str | None = None                  # PASO: Fino, Trocha, etc.

    # Nacimiento
    approximate_birth_date: date | None = None
    approximate_age_years: int | None = Field(default=None, ge=0)
    birth_date_is_approximate: bool = True
    birth_date_raw: str | None = None        # Texto original del Excel: "Oct. 15 de 2008"
    birth_place: str | None = None

    # Identificación externa
    registry_number: str | None = None       # # REGISTRO
    microchip: str | None = None             # MICROCHIP

    # Genealogía
    sire_name: str | None = None             # PADRE
    dam_name: str | None = None              # MADRE

    # Medidas actuales resumidas
    weight_kg: Decimal | None = Field(default=None, gt=0)
    height_m: Decimal | None = Field(default=None, gt=0)
    last_weight_at: date | None = None
    last_height_at: date | None = None

    # Operación
    is_active: bool = True
    is_available: bool = True
    operational_status: EquineOperationalStatus = EquineOperationalStatus.AVAILABLE
    availability_notes: str | None = None
    availability_reasons: str | None = None
    rest_until: datetime | None = None

    # Asignación
    max_rider_weight_kg: Decimal | None = Field(default=None, gt=0)
    experience_fit: EquineExperienceFit | None = EquineExperienceFit.ALL
    image_base64: str | None = None

    # Resumen calculado / mantenido por servicios
    last_service_at: datetime | None = None
    workload_last_7_days: int = Field(default=0, ge=0)

    # Trazabilidad de importación
    source_file: str | None = None
    source_sheet: str | None = None
    source_row_number: int | None = Field(default=None, ge=1)
    source_updated_at_label: str | None = None


class EquineUpdateSchema(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    inventory_number: int | None = Field(default=None, ge=1)

    species: EquineSpecies | None = None
    location_status: EquineLocationStatus | None = None
    location_notes: str | None = None

    breed: str | None = None
    sex: EquineSex | None = None
    coat_color: str | None = None
    gait: str | None = None

    approximate_birth_date: date | None = None
    approximate_age_years: int | None = Field(default=None, ge=0)
    birth_date_is_approximate: bool | None = None
    birth_date_raw: str | None = None
    birth_place: str | None = None

    registry_number: str | None = None
    microchip: str | None = None

    sire_name: str | None = None
    dam_name: str | None = None

    weight_kg: Decimal | None = Field(default=None, gt=0)
    height_m: Decimal | None = Field(default=None, gt=0)
    last_weight_at: date | None = None
    last_height_at: date | None = None

    is_active: bool | None = None
    is_available: bool | None = None
    operational_status: EquineOperationalStatus | None = None
    availability_notes: str | None = None
    availability_reasons: str | None = None
    rest_until: datetime | None = None

    max_rider_weight_kg: Decimal | None = Field(default=None, gt=0)
    experience_fit: EquineExperienceFit | None = None
    image_base64: str | None = None

    last_service_at: datetime | None = None
    workload_last_7_days: int | None = Field(default=None, ge=0)

    source_file: str | None = None
    source_sheet: str | None = None
    source_row_number: int | None = Field(default=None, ge=1)
    source_updated_at_label: str | None = None


class EquineResponseSchema(AuditMetadataSchema):
    id: str

    name: str
    inventory_number: int | None

    species: EquineSpecies
    location_status: EquineLocationStatus
    location_notes: str | None

    breed: str | None
    sex: EquineSex
    coat_color: str | None
    gait: str | None

    approximate_birth_date: date | None
    approximate_age_years: int | None
    birth_date_is_approximate: bool
    birth_date_raw: str | None
    birth_place: str | None

    registry_number: str | None
    microchip: str | None

    sire_name: str | None
    dam_name: str | None

    weight_kg: Decimal | None
    height_m: Decimal | None
    last_weight_at: date | None
    last_height_at: date | None

    is_active: bool
    is_available: bool
    operational_status: EquineOperationalStatus
    availability_notes: str | None
    availability_reasons: str | None
    rest_until: datetime | None

    max_rider_weight_kg: Decimal | None
    experience_fit: EquineExperienceFit | None
    image_base64: str | None = None

    last_service_at: datetime | None = None
    workload_last_7_days: int

    source_file: str | None
    source_sheet: str | None
    source_row_number: int | None
    source_updated_at_label: str | None


class EquineListItemSchema(AuditMetadataSchema):
    """Versión compacta para listados — omite genealogía, fechas de importación, trazabilidad.

    Incluye todos los campos necesarios para las cards de listado y el perfil
    del equino seleccionado, eliminando la necesidad de un request extra por detalle.
    """

    id: str
    name: str
    inventory_number: int | None
    species: EquineSpecies
    location_status: EquineLocationStatus = EquineLocationStatus.LA_JUANA
    location_notes: str | None = None
    breed: str | None
    sex: EquineSex
    coat_color: str | None
    gait: str | None
    weight_kg: Decimal | None
    height_m: Decimal | None
    is_active: bool
    is_available: bool
    operational_status: EquineOperationalStatus
    approximate_age_years: int | None = None
    max_rider_weight_kg: Decimal | None
    experience_fit: EquineExperienceFit | None
    image_base64: str | None = None
    last_service_at: datetime | None = None
    workload_last_7_days: int
    rest_until: datetime | None = None
    availability_reasons: str | None = None
    block_reason: str | None = None
    """Non-null when this equine is excluded from being assignable to a reservation."""


class EquineTimelineEntrySchema(BaseModel):
    """Entrada unificada del timeline: bitácora de servicio o evento de cuidado."""

    id: str
    source: Literal["service_log", "equine_event"]
    event_type: str
    happened_at: datetime
    title: str
    reservation_id: str | None = None
    assignment_id: str | None = None
    participant_id: str | None = None
    notes: str | None = None
    severity: str | None = None
    affects_availability: bool = False

File: packages/mobile_domain/lib/src/equines/equine.dart
import 'equine_experience_fit.dart';
import 'equine_operational_status.dart';
import '../gen/equine.dart' as gen;

/// Modelo de dominio completo para un equino (CV real).
class Equine {
  const Equine({
    required this.id,
    required this.name,
    this.inventoryNumber,
    this.species = 'mule',
    this.locationStatus = 'la_juana',
    this.locationNotes,
    this.breed,
    this.sex = 'unknown',
    this.coatColor,
    this.gait,
    this.approximateBirthDate,
    this.approximateAgeYears,
    this.birthDateIsApproximate = true,
    this.birthDateRaw,
    this.birthPlace,
    this.registryNumber,
    this.microchip,
    this.sireName,
    this.damName,
    this.weightKg,
    this.heightM,
    this.lastWeightAt,
    this.lastHeightAt,
    this.isActive = true,
    this.isAvailable = true,
    this.operationalStatus = EquineOperationalStatus.available,
    this.availabilityNotes,
    this.availabilityReasons,
    this.restUntil,
    this.maxRiderWeightKg,
    this.experienceFit,
    this.lastServiceAt,
    this.workloadLast7Days = 0,
    this.sourceFile,
    this.imageBase64,
    this.updatedAt,
  });

  /// Crea un [Equine] de dominio desde el modelo generado [gen.Equine].
  /// Convierte tipos: String→int?, String→double?, String→DateTime?, enum→String.
  factory Equine.fromGen(gen.Equine source) {
    return Equine(
      id: source.id,
      name: source.name,
      inventoryNumber: int.tryParse(source.inventoryNumber),
      species: source.species.value,
      locationStatus: source.locationStatus.value,
      locationNotes: source.locationNotes.isNotEmpty ? source.locationNotes : null,
      breed: source.breed.isNotEmpty ? source.breed : null,
      sex: source.sex.value,
      coatColor: source.coatColor.isNotEmpty ? source.coatColor : null,
      gait: source.gait.isNotEmpty ? source.gait : null,
      approximateBirthDate: source.approximateBirthDate.isNotEmpty
          ? source.approximateBirthDate
          : null,
      approximateAgeYears: int.tryParse(source.approximateAgeYears),
      birthDateIsApproximate: source.birthDateIsApproximate,
      birthDateRaw: source.birthDateRaw.isNotEmpty ? source.birthDateRaw : null,
      birthPlace: source.birthPlace.isNotEmpty ? source.birthPlace : null,
      registryNumber: source.registryNumber.isNotEmpty ? source.registryNumber : null,
      microchip: source.microchip.isNotEmpty ? source.microchip : null,
      sireName: source.sireName.isNotEmpty ? source.sireName : null,
      damName: source.damName.isNotEmpty ? source.damName : null,
      weightKg: double.tryParse(source.weightKg),
      heightM: double.tryParse(source.heightM),
      lastWeightAt: source.lastWeightAt.isNotEmpty ? source.lastWeightAt : null,
      lastHeightAt: source.lastHeightAt.isNotEmpty ? source.lastHeightAt : null,
      isActive: source.isActive,
      isAvailable: source.isAvailable,
      operationalStatus:
          EquineOperationalStatus.fromApi(source.operationalStatus.value),
      availabilityNotes:
          source.availabilityNotes.isNotEmpty ? source.availabilityNotes : null,
      availabilityReasons:
          source.availabilityReasons.isNotEmpty ? source.availabilityReasons : null,
      restUntil: source.restUntil.isNotEmpty
          ? DateTime.tryParse(source.restUntil)
          : null,
      maxRiderWeightKg: double.tryParse(source.maxRiderWeightKg),
      experienceFit: source.experienceFit.isNotEmpty
          ? EquineExperienceFit.fromApi(source.experienceFit)
          : null,
      lastServiceAt: source.lastServiceAt != null
          ? DateTime.tryParse(source.lastServiceAt!)
          : null,
      workloadLast7Days: source.workloadLast7Days,
      sourceFile: source.sourceFile.isNotEmpty ? source.sourceFile : null,
      imageBase64: source.imageBase64,
      updatedAt: source.updatedAt,
    );
  }

  final String id;
  final String name;
  final int? inventoryNumber;
  final String species;
  final String locationStatus;
  final String? locationNotes;
  final String? breed;
  final String sex;
  final String? coatColor;
  final String? gait;
  final String? approximateBirthDate;
  final int? approximateAgeYears;
  final bool birthDateIsApproximate;
  final String? birthDateRaw;
  final String? birthPlace;
  final String? registryNumber;
  final String? microchip;
  final String? sireName;
  final String? damName;
  final double? weightKg;
  final double? heightM;
  final String? lastWeightAt;
  final String? lastHeightAt;
  final bool isActive;
  final bool isAvailable;
  final EquineOperationalStatus operationalStatus;
  final String? availabilityNotes;
  final String? availabilityReasons;
  final DateTime? restUntil;
  final double? maxRiderWeightKg;
  final EquineExperienceFit? experienceFit;
  final DateTime? lastServiceAt;
  final int workloadLast7Days;
  final String? sourceFile;
  final String? imageBase64;
  final DateTime? updatedAt;
}
